### ISSUE TITLE
Build error on aarch64

### ISSUE THREAD

**Issue by sxyazi** [jstkdng/ueberzugpp#96]
I got a build error when I run `paru -S ueberzugpp`:

```
-- The C compiler identification is GNU 12.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/sbin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring with CMAKE_SYSTEM_PROCESSOR aarch64
-- Configuring done (0.1s)
-- Generating done (0.0s)
-- Build files have been written to: /home/dev/.cache/paru/clone/turbo-base64/src/build
[ 75%] Building C object CMakeFiles/_b64_scalar.dir/turbob64d.c.o
[ 75%] Building C object CMakeFiles/_b64_v128.dir/turbob64v128.c.o
[ 75%] Building C object CMakeFiles/_b64_scalar.dir/turbob64c.c.o
/home/dev/.cache/paru/clone/turbo-base64/src/Turbo-Base64-2023.04/turbob64v128.c:135:3: error: expected identifier or '(' before 'return'
  135 |   return (op - out) + rc;
      |   ^~~~~~
/home/dev/.cache/paru/clone/turbo-base64/src/Turbo-Base64-2023.04/turbob64v128.c:136:1: error: expected identifier or '(' before '}' token
  136 | }
      | ^
[ 75%] Built target _b64_scalar
make[2]: *** [CMakeFiles/_b64_v128.dir/build.make:76: CMakeFiles/_b64_v128.dir/turbob64v128.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:113: CMakeFiles/_b64_v128.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
==> ERROR: A failure occurred in build().
    Aborting...
error: failed to build 'turbo-base64-2023.04-1':
error: can't build ueberzugpp-2.8.9-1, deps not satisfied: turbo-base64
error: packages failed to build: turbo-base64-2023.04-1  ueberzugpp-2.8.9-1
```

The error comes from the upstream of Turbo-Base64 package, and a corresponding patch is there: https://github.com/powturbo/Turbo-Base64/pull/21

**Comment by justchokingaround** [comment]
also doesn't compile with the wayland flag:

```
 17%] Building C object CMakeFiles/ueberzug.dir/wayland-xdg-shell-protocol.c.o
[ 20%] Building CXX object CMakeFiles/ueberzug.dir/src/canvas/wayland/wayland.cpp.o
In file included from /usr/include/c++/13.1.1/bits/stl_tempbuf.h:61,
                 from /usr/include/c++/13.1.1/memory:66,
                 from /home/chokerman/dev/ueberzugpp/include/canvas.hpp:20,
                 from /home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.hpp:20,
                 from /home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.cpp:17:
/usr/include/c++/13.1.1/bits/stl_construct.h: In instantiation of ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}]’:
/usr/include/c++/13.1.1/bits/alloc_traits.h:660:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; allocator_type = std::allocator<void>]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}; _NonArray<_Tp> = WaylandShmWindow]’
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.cpp:158:48:   required from here
/usr/include/c++/13.1.1/bits/stl_construct.h:115:28: error: no matching function for call to ‘construct_at(WaylandShmWindow*&, wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image>, std::shared_ptr<WaylandConfig>&)’
  115 |           std::construct_at(__p, std::forward<_Args>(__args)...);
      |           ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/13.1.1/bits/stl_construct.h:94:5: note: candidate: ‘template<class _Tp, class ... _Args> constexpr decltype (::new(void*(0)) _Tp) std::construct_at(_Tp*, _Args&& ...)’
   94 |     construct_at(_Tp* __location, _Args&&... __args)
      |     ^~~~~~~~~~~~
/usr/include/c++/13.1.1/bits/stl_construct.h:94:5: note:   template argument deduction/substitution failed:
/usr/include/c++/13.1.1/bits/stl_construct.h: In substitution of ‘template<class _Tp, class ... _Args> constexpr decltype (::new(void*(0)) _Tp) std::construct_at(_Tp*, _Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}]’:
/usr/include/c++/13.1.1/bits/stl_construct.h:115:21:   required from ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}]’
/usr/include/c++/13.1.1/bits/alloc_traits.h:660:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; allocator_type = std::allocator<void>]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}; _NonArray<_Tp> = WaylandShmWindow]’
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.cpp:158:48:   required from here
/usr/include/c++/13.1.1/bits/stl_construct.h:96:17: error: no matching function for call to ‘WaylandShmWindow::WaylandShmWindow(wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image>, std::shared_ptr<WaylandConfig>&)’
   96 |     -> decltype(::new((void*)0) _Tp(std::declval<_Args>()...))
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.cpp:31:
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/window/waylandshm.hpp:37:5: note: candidate: ‘WaylandShmWindow::WaylandShmWindow(wl_compositor*, wl_shm*, xdg_wm_base*, std::unique_ptr<Image>, std::shared_ptr<WaylandConfig>, XdgStructAgg&)’
   37 |     WaylandShmWindow(struct wl_compositor *compositor, struct wl_shm *wl_shm,
      |     ^~~~~~~~~~~~~~~~
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/window/waylandshm.hpp:37:5: note:   candidate expects 6 arguments, 5 provided
/usr/include/c++/13.1.1/bits/stl_construct.h: In instantiation of ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}]’:
/usr/include/c++/13.1.1/bits/alloc_traits.h:660:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; allocator_type = std::allocator<void>]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = WaylandShmWindow; _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image, std::default_delete<Image> >, std::shared_ptr<WaylandConfig>&}; _Tp = WaylandShmWindow]’
/usr/include/c++/13.1.1/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = WaylandShmWindow; _Args = {wl_compositor*&, wl_shm*&, xdg_wm_base*&, unique_ptr<Image, default_delete<Image> >, shared_ptr<WaylandConfig>&}; _NonArray<_Tp> = WaylandShmWindow]’
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/wayland.cpp:158:48:   required from here
/usr/include/c++/13.1.1/bits/stl_construct.h:119:7: error: no matching function for call to ‘WaylandShmWindow::WaylandShmWindow(wl_compositor*&, wl_shm*&, xdg_wm_base*&, std::unique_ptr<Image>, std::shared_ptr<WaylandConfig>&)’
  119 |       ::new((void*)__p) _Tp(std::forward<_Args>(__args)...);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/window/waylandshm.hpp:37:5: note: candidate: ‘WaylandShmWindow::WaylandShmWindow(wl_compositor*, wl_shm*, xdg_wm_base*, std::unique_ptr<Image>, std::shared_ptr<WaylandConfig>, XdgStructAgg&)’
   37 |     WaylandShmWindow(struct wl_compositor *compositor, struct wl_shm *wl_shm,
      |     ^~~~~~~~~~~~~~~~
/home/chokerman/dev/ueberzugpp/src/canvas/wayland/window/waylandshm.hpp:37:5: note:   candidate expects 6 arguments, 5 provided
make[2]: *** [CMakeFiles/ueberzug.dir/build.make:141: CMakeFiles/ueberzug.dir/src/canvas/wayland/wayland.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/ueberzug.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Comment by jstkdng** [comment]
@justchokingaround are you trying to compile from master?

**Comment by justchokingaround** [comment]
yes

**Comment by jstkdng** [comment]
ah, I get the issue as well if opengl is disabled. Should be fixed now.

@sxyazi I'll add the patch once I have it working

**Comment by justchokingaround** [comment]
it now compiles fine, however it fails to render some images, both for wayland and sixel as output:
![image](https://github.com/jstkdng/ueberzugpp/assets/44473782/b61293e0-75b8-4a17-aabf-a08a66fa7092)

should i open a separate issue for this?

**Comment by jstkdng** [comment]
yeah, open a separate issue.

**Comment by jstkdng** [comment]
hey @powturbo , would you be able to create a new release? for some reason I'm unable to patch correctly the source, could be because of the windows line endings.

**Comment by powturbo** [comment]
done!

**Comment by jstkdng** [comment]
uhh, you haven't made a new release, 2023.04 is still the latest one

**Comment by powturbo** [comment]
I'm doing normally only git update. 
I've created a new release 2023.08 now.

**Comment by jstkdng** [comment]
@powturbo nice, but I think the issue is still there, didn't the pr linked here fix it?

```
-- The C compiler identification is GNU 12.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring with CMAKE_SYSTEM_PROCESSOR aarch64
-- Configuring done (0.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/alarm/turbo-base64/src/build
[ 75%] Building C object CMakeFiles/_b64_v128.dir/turbob64v128.c.o
[ 75%] Building C object CMakeFiles/_b64_scalar.dir/turbob64d.c.o
[ 75%] Building C object CMakeFiles/_b64_scalar.dir/turbob64c.c.o
In file included from /home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64v128.c:49:
/home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64v128.c: In function ‘tb64v128enc’:
/home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64_.h:71:32: error: ‘out_’ undeclared (first use in this function); did you mean ‘out’?
   71 | #define EXTAIL(_n_) for(; op < out_-4; op += 4, ip += 3) { unsigned _u = BSWAP32(ctou32(ip)); stou32(op, XU32(_u)); } ETAIL()
      |                                ^~~~
/home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64v128.c:183:3: note: in expansion of macro ‘EXTAIL’
  183 |   EXTAIL();
      |   ^~~~~~
/home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64_.h:71:32: note: each undeclared identifier is reported only once for each function it appears in
   71 | #define EXTAIL(_n_) for(; op < out_-4; op += 4, ip += 3) { unsigned _u = BSWAP32(ctou32(ip)); stou32(op, XU32(_u)); } ETAIL()
      |                                ^~~~
/home/alarm/turbo-base64/src/Turbo-Base64-2023.08/turbob64v128.c:183:3: note: in expansion of macro ‘EXTAIL’
  183 |   EXTAIL();
      |   ^~~~~~
make[2]: *** [CMakeFiles/_b64_v128.dir/build.make:76: CMakeFiles/_b64_v128.dir/turbob64v128.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:113: CMakeFiles/_b64_v128.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 75%] Built target _b64_scalar
```

**Comment by sxyazi** [comment]
Weird, the typo of `out_` in `turbob64_.h:71:32` should be fixed to `out` by https://github.com/powturbo/Turbo-Base64/pull/21/files#diff-1b8120c79e828e21df9f259734b63c77e3378c656a3fe974798297346ecf9754

**Comment by powturbo** [comment]
Unfortunately I've actually, no arm64 system for testing.
I've now made a change in the arm neon section (out_ was not declared) and created a new release.
Can you please try again?

**Comment by sxyazi** [comment]
Yeah, how can I test it?

**Comment by powturbo** [comment]
Can you please try again. I'll check this tomorow with cross compiling on linux.

**Comment by jstkdng** [comment]
yeah, it works now, thanks.

**Comment by jstkdng** [comment]
@sxyazi I've updated the `turbo-base64` package and tested on my rpi, should be working now.

**Comment by sxyazi** [comment]
It works now like a charm, thank you all!
